src/main/flight/imu.c

```diff
@@ -261,150 +243,158 @@ static void imuCheckAndResetOrientationQuaternion(const float ax, const float ay
-        ex = (ay * rMat[2][2] - az * rMat[2][1]);
-        ey = (az * rMat[2][0] - ax * rMat[2][2]);
-        ez = (ax * rMat[2][1] - ay * rMat[2][0]);
+        vectorNormalize(&vAcc, accBF);
+        vectorCrossProduct(&vErr, &vAcc, &vEstGravity);
 
         // Compute and apply integral feedback if enabled
         if (imuRuntimeConfig.dcm_ki_acc > 0.0f) {
             // Stop integrating if spinning beyond the certain limit
             if (spin_rate_sq < sq(DEGREES_TO_RADIANS(SPIN_RATE_LIMIT))) {
-                integralAccX += imuRuntimeConfig.dcm_ki_acc * ex * dt;    // integral error scaled by Ki
-                integralAccY += imuRuntimeConfig.dcm_ki_acc * ey * dt;
-                integralAccZ += imuRuntimeConfig.dcm_ki_acc * ez * dt;
+                t_fp_vector vTmp;
 
-                gx += integralAccX;
-                gy += integralAccY;
-                gz += integralAccZ;
+                // integral error scaled by Ki
+                vectorScale(&vTmp, &vErr, imuRuntimeConfig.dcm_ki_acc * dt);
+                vectorAdd(&vIntegralAcc, &vIntegralAcc, &vTmp);
+
+                // apply integral feedback
+                vectorAdd(&vRotation, &vRotation, &vIntegralAcc);
             }
         }
 
         // Calculate kP gain and apply proportional feedback
-        gx += kpAcc * ex;
-        gy += kpAcc * ey;
-        gz += kpAcc * ez;
+        vectorScale(&vErr, &vErr, imuRuntimeConfig.dcm_kp_acc * imuGetPGainScaleFactor());
+        vectorAdd(&vRotation, &vRotation, &vErr);
     }
 
     // Integrate rate of change of quaternion
-    gx *= (0.5f * dt);
-    gy *= (0.5f * dt);
-    gz *= (0.5f * dt);
-
-    const float qa = q0;
-    const float qb = q1;
-    const float qc = q2;
-    q0 += (-qb * gx - qc * gy - q3 * gz);
-    q1 += (qa * gx + qc * gz - q3 * gy);
-    q2 += (qa * gy - qb * gz + q3 * gx);
-    q3 += (qa * gz + qb * gy - qc * gx);
-
-    // Normalise quaternion
-    const float quatRecipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3);
-    q0 *= quatRecipNorm;
-    q1 *= quatRecipNorm;
-    q2 *= quatRecipNorm;
-    q3 *= quatRecipNorm;
+    t_fp_vector vTheta;
+    fpQuaternion_t deltaQ;
+
+    vectorScale(&vTheta, &vRotation, 0.5f * dt);
+    quaternionInitFromVector(&deltaQ, &vTheta);
+    const float thetaMagnitudeSq = vectorNormSquared(&vTheta);
+
+    // Calculate quaternion delta:
+    // Theta is a axis/angle rotation. Direction of a vector is axis, magnitude is angle/2.
+    // Proper quaternion from axis/angle involves computing sin/cos, but the formula becomes numerically unstable as Theta approaches zero.
+    // For near-zero cases we use the first 3 terms of the Taylor series expansion for sin/cos. We check if fourth term is less than machine precision -
+    // then we can safely use the "low angle" approximated version without loss of accuracy.
+    if (thetaMagnitudeSq * thetaMagnitudeSq / 24.0f < 1e-6f) {
```

Comment:
you can sqrt both sides ... -> `thetaMagnitudeSq < sqrt(24*1e-6)`

Reply:
`sqrtf` if more expensive than multiplication

Reply:
But square root of constant is free ...

Reply:
Oops 😄